DAOS-6428 daos: cleanup filesystem copy

dsikich opened the pull request:
Address comments in PR #4193:
    * use gurt macros for string functions
    * do not update source permissions
    * remove assert after fs_copy_hdlr
    * close files on error in fs_copy
    * close dst if umount fails on src
    * abstract src and dst parsing to function
    * use gurt macros to free memory
    * free dfs_path
    * fix alignment
    * do not define var in middle of code
    * print errno on error in file_write
    * update docs

Signed-off-by: Danielle Sikich <danielle.sikich@intel.com>

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2657,159 +2655,101 @@ static int fs_copy_parse_uuids(char *str,
-{
-	/* parse src and dst strings
+fs_copy_parse(struct file_dfs *file_dfs,
+		struct duns_attr_t *dattr,
```
```suggestion
	      struct duns_attr_t *dattr,
```

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/1/execution/node/60/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

dsikich commented:
@daltonbohning @johannlombardi @mchaarawi this PR addresses comments in PR #4193

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/3/execution/node/290/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/3/execution/node/280/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/3/execution/node/289/log

daosbuild1 commented:
Test stage Build on Ubuntu 20.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/3/execution/node/298/log

daosbuild1 commented:
Test stage Build on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/3/execution/node/319/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/3/execution/node/320/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented on doc/man/man8/daos.8:
```diff
@@ -196,6 +199,17 @@ The \fBRESOURCE\fRs, respective \fBCOMMAND\fRs and \fBOPTION\fRs supported by \f
+.TP
+.I filesystem \fR(\fIfs\fR) \fBOPTION\fRs:
+.br
+	  \fB--src=\fRSTR     src=daos://<pool/cont> | <path>
+.br
+	  \fB--dst=\fRSTR     dst=daos://<pool/cont> | <path>
```
@dsikich Looking at this with fresh eyes...
Maybe we should clarify that a daos src/dst can have a path as well?
```suggestion
	  \fB--src=\fRSTR     src=daos://<pool/cont>[/<path>] | <path>
.br
	  \fB--dst=\fRSTR     dst=daos://<pool/cont>[/<path>] | <path>
```

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1613,19 +1613,19 @@ parse_filename_dfs(const char *path, char **_obj_name, char **_cont_name)
 	}
-	f1 = strdup(path);
+	D_STRNDUP(f1, path, strlen(path) + 1);
 	if (f1 == NULL) {
 		rc = -ENOMEM;
 		goto out;
 	}
 
-	f2 = strdup(path);
+	D_STRNDUP(f2, path, strlen(path) + 1);
```
@dsikich Maybe calculate `strlen(path)` once and use it in both places?
For what it's worth, this function will be refactored when the DFS wrappers are created as part of [DAOS-6098](https://jira.hpdd.intel.com/browse/DAOS-6098)

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2657,159 +2655,101 @@ static int fs_copy_parse_uuids(char *str,
-			fprintf(stderr, "cannot parse daos dst type format, "
-				"please use:\n"
-				"\t--dst=daos://<pool/cont>\n");
+				"\t--<src | dst>=daos://<pool/cont>\n");
```
@dsikich Similar to previous. Should we have `<path>` here for a daos subset?

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2657,159 +2655,101 @@ static int fs_copy_parse_uuids(char *str,
-		fprintf(stderr, "cannot parse src format, please use:\n"
-				"--src=<daos>://<pool/cont> | <path>\n"
+		fprintf(stderr, "cannot parse arg format, please use:\n"
+				"--<src | dst>=<daos>://<pool/cont> | <path>\n"
```
@dsikich Similar to previous. Should we have `<path>` here for a daos subset?

daltonbohning reviewed (approved):
For what it's worth, my local branch with basic `daos filesytem copy` tests is passing.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mchaarawi commented on doc/man/man8/daos.8:
```diff
@@ -28,7 +28,8 @@ daos \- command to manage DAOS pool/container/object
 .B daos
 can be used to manage/query a pool content, create/query/manage/destroy a
-container inside a pool or query/manage an object inside a container.
+container inside a pool, copy data to/from a POSIX container to POSIX,
```
copy data to/from a DAOS POSIX container to a POSIX filesystem

mchaarawi commented on doc/man/man8/daos.8:
```diff
@@ -38,6 +39,8 @@ The \fBRESOURCE\fRs, respective \fBCOMMAND\fRs and \fBOPTION\fRs supported by \f
 .br
 	  \fBcontainer \fR(\fBcont\fR) container in a pool
+.br
+	  \fBfilesystem \fR(\fBfs\fR) POSIX path or POSIX container
```
POSIX filesystem path

dsikich commented:
@daltonbohning @mchaarawi sorry, I updated the docs and needed to restart the Jenkins so it dismissed your approvals. I just updated the working in the docs @mchaarawi requested. It was not allowing me to just restart the Jenkins from the web interface.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mchaarawi commented on doc/man/man8/daos.8:
```diff
@@ -40,7 +40,7 @@ The \fBRESOURCE\fRs, respective \fBCOMMAND\fRs and \fBOPTION\fRs supported by \f
 	  \fBcontainer \fR(\fBcont\fR) container in a pool
 .br
-	  \fBfilesystem \fR(\fBfs\fR) POSIX path or POSIX container
+	  \fBfilesystem \fR(\fBfs\fR) POSIX path filesystem path or POSIX container
```
typo: extra path

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

dsikich commented:
@daltonbohning sorry to spam you, I mean to request your review on my other PR.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -1607,25 +1607,26 @@ parse_filename_dfs(const char *path, char **_obj_name, char **_cont_name)
 	char	*f2 = NULL;
 	char	*fname = NULL;
 	char	*cont_name = NULL;
+	int	path_len = strlen(path) + 1;
```
We should be avoiding the use of strlen() wherever possible, in this case path seems to come from a stack variable in fs_copy() which is a known size so there's no reason not to pass in the size, however there is another issue here which is that MAX_FILENAME(256) is nowhere near long enough.

char filename[MAX_FILENAME] should be replaced by D_ASPRINTF(), and tracking of len.

dsikich replied:
@ashleypittman this would require changing every single dfs wrapper since parse_filename_dfs is used in just about every single one. I do think improving this function can be looked at under https://jira.hpdd.intel.com/browse/DAOS-6449 since we are refactoring the dfs wrappers under that ticket, and moving them to another location.

dsikich replied:
@ashleypittman I updated the filenames used in fs_copy to use D_ASPRINTF since it is recursively walking the file tree, and you are right 256 was not long enough.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -1642,7 +1643,7 @@ parse_filename_dfs(const char *path, char **_obj_name, char **_cont_name)
 
 		if (strcmp(cont_name, ".") == 0) {
-			cont_name = strdup(cwd);
+			D_STRNDUP(cont_name, cwd, strlen(cwd) + 1);
```
The length of the buffer is known here, so use it.  strlen() gives you no protection at all.
```suggestion
			D_STRNDUP(cont_name, cwd, 1024);
```

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -1667,33 +1668,33 @@ parse_filename_dfs(const char *path, char **_obj_name, char **_cont_name)
-			free(f1);
-		if (f2)
-			free(f2);
+	if (f1)
+		D_FREE(f1);
+	if (f2)
+		D_FREE(f2);
```
```suggestion
	D_FREE(f1);
	D_FREE(f2);
```

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -1728,8 +1729,8 @@ file_write(struct file_dfs *file_dfs, char *file,
 	if (num_bytes_written < 0) {
-		fprintf(stderr, "write error on %s type=%d\n",
-			file, file_dfs->type);
+		fprintf(stderr, "write error on %s type=%d errno=%s\n",
```
This is wrong if writing to daos, errno will not be set in that case.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -1895,8 +1896,8 @@ file_mkdir(struct file_dfs *file_dfs, const char *dir, mode_t *mode)
-	int rc;
-	struct fs_copy_dirent *dirp = calloc(1, sizeof(*dirp));
+	int	rc = 0;
+	struct	fs_copy_dirent *dirp = calloc(1, sizeof(*dirp));
```
There are a couple of uses in calloc() in this code, but this is the only one that I can comment on.
```suggestion
	struct	fs_copy_dirent *dirp;
	
	 D_ALLOC_PTR(dirp);
```

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -1906,7 +1907,7 @@ opendir_dfs(struct file_dfs *file_dfs, const char *dir)
 		fprintf(stderr, "dfs_lookup %s failed\n", dir);
 		errno = rc;
-		free(dirp);
+		D_FREE(dirp);
```
D_FREE() sets dirp to NULL so you don't need the following line.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -1931,8 +1932,8 @@ file_opendir(struct file_dfs *file_dfs, const char *dir)
-	int rc;
-	struct fs_copy_dirent *dirp = (struct fs_copy_dirent *)_dirp;
+	int	rc = 0;
+	struct	fs_copy_dirent *dirp = (struct fs_copy_dirent *)_dirp;
```
This function is static, why is there a need to cast one of the input variables?

dsikich replied:
these dfs wrappers are wrapped by file_ functions that call either the POSIX IO function or the DAOS IO function. The file_ function takes in a DIR *dirp, so these wrappers have similar arguments across POSIX, DAOS. Otherwise, fs_copy_dirent would have to be passed in even on calls for POSIX directories. For what its worth, these wrappers are going to be moved into DFS, and refactored there by @daltonbohning.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2239,9 +2240,11 @@ fs_copy(struct file_dfs *src_file_dfs,
+	DIR	*src_dir = NULL;
+	struct	stat st_dir_name;
+	char	filename[MAX_FILENAME];
+	char	dst_filename[MAX_FILENAME];
```
Alignment.

dsikich replied:
@ashleypittman I'm not seeing this alignment issue here?

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2467,12 +2457,20 @@ fs_copy(struct file_dfs *src_file_dfs,
+	 * otherwise always close it before returning
+	 */
+	if (S_ISDIR(st_dir_name.st_mode)) {
+		if (file_closedir(src_file_dfs, src_dir))
+			fprintf(stderr, "Could not close '%s': %s\n",
+				dir_name, strerror(errno));
```
Again, errno is not the right thing here.
```suggestion
                int err = file_closedir(src_file_dfs, src_dir);
                
                if (err != 0)
			fprintf(stderr, "Could not close '%s': %s\n",
				dir_name, strerror(err));
```

dsikich replied:
@ashleypittman the rc is the errno, the file_closedir call returns an errno as its rc, so I changed errno to rc. dfs_closedir returns an errno as its rc, and the posix call sets errno, which is returned from file_closedir.

ashleypittman replied:
file_closedir() does return the errno, but the log message is using errno again, which will be unset in the DAOS case.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2657,159 +2656,103 @@ static int fs_copy_parse_uuids(char *str,
-		src_type = "daos";
-		/* start of src pool/cont str */
-		src += 7;
+	if (strncmp(arg, "daos://", 7) == 0) {
```
I'm not familiar with what this is doing, but I do wonder if duns_resolve_path() would help here, it already parses daos://<uuid>/<uuid> which from the comments is what whis is doing.

dsikich replied:
@ashleypittman I don't think that was available when I started writing this, but I moved to using the duns_resolve_path call to get the rel_path, so that should resolve most of your comments on the parsing below.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2657,159 +2656,103 @@ static int fs_copy_parse_uuids(char *str,
-		int src_pool_cont_len = (strlen(src_pool) +
-					strlen(src_cont) + 1);
+		len = strlen(*dfs_path);
+		pool_cont_len = (strlen(pool) +
```
If pool and cont are uuids then it's a known length.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2657,159 +2656,103 @@ static int fs_copy_parse_uuids(char *str,
-			*src_dfs_path += strlen(src_pool) +
-					strlen(src_cont) + 1;
-			*src_str = strdup(*src_dfs_path);
+			*dfs_path += strlen(pool) + strlen(cont) + 1;
```
You've literally just called strlen(pool) five lines above.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2657,159 +2656,103 @@ static int fs_copy_parse_uuids(char *str,
+			 * cont subset path is where the copy starts
+			 */
+			D_STRNDUP(*arg_str, *dfs_path,
+				  strlen(*dfs_path) + 1);
```
And this, 11 lines above.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2657,159 +2656,103 @@ static int fs_copy_parse_uuids(char *str,
+			 * since it is no longer needed
+			 */
+			*dfs_path -= strlen(pool) + strlen(cont) + 1;
+			if (*dfs_path != NULL)
```
if *dfs_path is NULL here then something very weird is going on.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2834,18 +2777,28 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	char			*dname = NULL;
+	char			dst_dir[MAX_FILENAME];
+	int			path_length = 0;
+	mode_t			tmp_mode_dir;
```
Why not set this here, it looks to always be the same value?  Alternatively just use a define.

ashleypittman reviewed (commented):
This PR is a big step forwards but more work is still needed on the topic.  I also wonder if this code is in the correct src file, or if it would be better placed elsewhere in the src tree.

daltonbohning commented:
@ashleypittman For what it's worth, these `dfs` wrappers and the parse function will be abstracted to use the wrappers defined as part of [DAOS-6449](https://jira.hpdd.intel.com/browse/DAOS-6449), so I do think that at least some of the changes here will be handled by DAOS-6449.

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2729,186 +2701,41 @@ fs_copy_disconnect(struct fs_copy_args *fa,
-				"\ttype is daos, only specified if "
-				"pool/cont used\n");
-		D_GOTO(out, rc = EINVAL);
+		/* not a DAOS path, set type to POSIX, 
```
(style)  trailing whitespace

daosbuild1 replied:
```suggestion
		/* not a DAOS path, set type to POSIX,
```

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4224/11/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/11/execution/node/66/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build on Ubuntu 20.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/12/execution/node/332/log

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2337,9 +2336,14 @@ fs_copy(struct file_dfs *src_file_dfs,
+	DIR	*src_dir = NULL;
+	struct	stat st_dir_name;
+	char	*filename = NULL;
+	char	*dst_filename= NULL;
```
(style)  spaces required around that '=' (ctx:VxW)

daosbuild1 replied:
```suggestion
	char	*dst_filename = NULL;
```

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4224/13/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4224/13/execution/node/66/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2398,15 +2392,8 @@ fs_copy(struct file_dfs *src_file_dfs,
-			fprintf(stderr, "Path length is too long.\n");
-			D_GOTO(out, rc);
-		}
+		D_ASPRINTF(dst_filename, "%s/%s", fs_dst_prefix,
```
You need to check dst_filename != NULL after this.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2376,17 +2380,7 @@ fs_copy(struct file_dfs *src_file_dfs,
-			fprintf(stderr, "Path length is too long.\n");
-			D_GOTO(out, rc);
-		}
+		D_ASPRINTF(filename, "%s/%s", dir_name, d_name);
```
And here.

ashleypittman commented on src/utils/daos_hdlr.c:
```diff
@@ -2492,29 +2473,12 @@ fs_copy(struct file_dfs *src_file_dfs,
-						"long on destination.\n");
-					D_GOTO(out, rc);
-				}
+				D_ASPRINTF(next_path, "%s", filename);
```
D_STRNDUP() for these two, but you need to check the pointer afterwards.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ashleypittman reviewed (approved):
More work is needed in this area but this is a big step forward.
